Fix to accept parameters out of init scope

disktnk opened the pull request:
fixes #80

shinh commented on onnx_chainer/export.py:
```diff
@@ -259,6 +266,16 @@ def export(model, args, filename=None, export_params=True,
+    if out_scoped_param_names:
+        for name in out_scoped_param_names:
+            param = o.inputs[name]
+            initializers.append(convert_parameter(param))
```
JFYI: currently, ONNX-chainer is using input+initializer instead of `Constant` for this kind of constant values. However, I think `Constant` is more appropriate and input+initializer should be only for network parameters. This is OK, but it'd be great if we can eventually fix it.

disktnk replied:
agree, I issued #82

shinh replied:
Thanks!

disktnk commented on onnx_chainer/export.py:
```diff
@@ -259,6 +266,16 @@ def export(model, args, filename=None, export_params=True,
+        for name in out_scoped_param_names:
+            param = o.inputs[name]
+            initializers.append(convert_parameter(param))
+            param_shape = (1,) if param.shape == () else param.shape
```
fixed

shinh replied:
I still see this line? Ah, you have copied this line from the one for parameters?

disktnk commented on onnx_chainer/export.py:
```diff
@@ -259,6 +266,16 @@ def export(model, args, filename=None, export_params=True,
 
+    out_scoped_param_names = set(o.inputs.keys()) - param_names -\
+        network_input_names
+    if out_scoped_param_names:
```
definitely

disktnk commented on onnx_chainer/export.py:
```diff
@@ -259,6 +266,16 @@ def export(model, args, filename=None, export_params=True,
             outputs.grad = numpy.ones_like(outputs.array)
             outputs.backward()
 
+    out_scoped_param_names = set(o.inputs.keys()) - param_names -\
```
renamed, thx

shinh commented on onnx_chainer/export.py:
```diff
@@ -232,14 +235,18 @@ def export(model, args, filename=None, export_params=True,
     for param in model.params():
+        param_names.add(str(id(param)))
         initializers.append(convert_parameter(param))
         param_shape = (1,) if param.shape == () else param.shape
```
I think this line should be removed, too.

disktnk commented:
sorry, this comments https://github.com/chainer/onnx-chainer/pull/81#discussion_r255380624 is wrong, I removed the line but uncommitted.

However, "`param_shape = (1,) if param.shape == () else param.shape`" is required because `convert_parameter` function takes care about it, so must to be matched the shape.

so I fixed them as refactoring

1. share tensor made by `convert_parameter` function and remove "`param_shape = (1,) if param.shape == () else param.shape`" in https://github.com/chainer/onnx-chainer/pull/81/commits/172d85989fce5e937983354f81cfd86eed9cbbfb
2. remove the `array = array[None]` when the value's shape is empty in https://github.com/chainer/onnx-chainer/pull/81/commits/a1e85d60ffe972ae24c363265a93325051ea57a7

shinh commented on onnx_chainer/export.py:
```diff
@@ -43,8 +43,6 @@ def convert_parameter(parameter):
             'The type of parameter is unknown. It should be either Parameter '
             'or Variable or ndarray, but the type was {}.'.format(
                 type(parameter)))
-    if array.shape == ():
```
Nice! I haven't realized this and I was wondering why `param_shape = (1,)` thing is necessary.
